libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -0,0 +1,3082 @@
+    suffix :: forall as'. List (Term univ) as' -> m (List Value as')
+    suffix Nil = pure Nil
+    suffix (Lit l :> ts) = (Value l :>) <$> suffix ts
+    suffix (_ :> _) = fatalError ["toCtxList with too many holes"]
+
+------------------------------------------------------------------------
+-- Semantics of `Term` and `Pred`
+------------------------------------------------------------------------
+
+runTerm :: MonadGenError m => Env -> Term univ a -> m a
+runTerm env = \case
+  Lit a -> pure a
+  V v -> findEnv env v
+  App f ts -> do
+    vs <- mapMList (fmap Identity . runTerm env) ts
+    pure $ uncurryList_ runIdentity (sem f) vs
+
+checkPred :: forall univ m. (FunctionLike univ, MonadGenError m) => Env -> Pred univ -> m Bool
+checkPred env = \case
+  Assert [] t -> runTerm env t
+  Assert es t -> explain ("assert:" : map ("  " ++) es) $ runTerm env t
+  Reify t f (x :-> p) -> do
+    val <- runTerm env t
+    checkPred (extendEnv x (f val) env) p
+  ForAll t (x :-> p) -> do
+    set <- runTerm env t
+    and
+      <$> sequence
+        [ checkPred env' p
+        | v <- forAllToList set
+        , let env' = extendEnv x v env
+        ]
+  Case t bs -> do
+    v <- runTerm env t
+    runCaseOn v bs (\x val ps -> checkPred (extendEnv x val env) ps)
+  TruePred -> pure True
+  FalsePred es -> explain es $ pure False
+  DependsOn {} -> pure True
+  Block ps -> checkPreds env ps
+  Let t (x :-> p) -> do
+    val <- runTerm env t
+    checkPred (extendEnv x val env) p
+  Exists k (x :-> p) -> do
+    a <- runGE . k $ errorGE . explain ["checkPred: Exists"] . runTerm env
+    checkPred (extendEnv x a env) p
+
+checkPreds :: (MonadGenError m, Traversable t, FunctionLike univ) => Env -> t (Pred univ) -> m Bool
+checkPreds env ps = and <$> mapM (checkPred env) ps
+
+checkPredPure :: FunctionLike univ => Env -> Pred univ -> Bool
+checkPredPure env p = fromGE (const False) $ checkPred env p
+
+runCaseOn ::
+  SumOver as ->
+  List (Binder univ) as ->
+  (forall a. HasSpec univ a => Var a -> a -> Pred univ -> r) ->
+  r
+runCaseOn _ Nil _ = error "The impossible happened in runCaseOn"
+runCaseOn a ((x :-> ps) :> Nil) f = f x a ps
+runCaseOn s ((x :-> ps) :> bs@(_ :> _)) f = case s of
+  SumLeft a -> f x a ps
+  SumRight a -> runCaseOn a bs f
+
+------------------------------------------------------------------------
+-- Specs
+------------------------------------------------------------------------
+
+-- NOTE: in the future one might consider using sets when a type
+-- has `Ord` here. Beware, this means that one needs to have a check
+-- for instances of Ord at runtime!
+type OrdSet a = [a]
+
+-- | A `Spec univ a` denotes a set of `a`s
+data Spec univ a where
+  -- | Elements of a known set
+  MemberSpec ::
+    -- | It must be an element of this set
+    OrdSet a ->
+    Spec univ a
+  -- | The empty set
+  ErrorSpec ::
+    [String] ->
+    Spec univ a
+  -- | The set described by some predicates
+  -- over the bound variable
+  SuspendedSpec ::
```

Comment:
I'm not really sure what this does.

Reply:
C.f. `checkPred` for the semantics, basically the variable ranges over the values denoted by the spec and the variable is subject to the `Pred`

Reply:
Sorry, not `checkPred` obviously... I mean `conformsToSpecM`